src/SpectatorView.Unity/Assets/SpectatorView/Scripts/StateSynchronization/AssetCache.cs

```diff
@@ -277,35 +404,129 @@ public override void ClearAssetCache()
         public override void UpdateAssetCache()
         {
 #if UNITY_EDITOR
-            Dictionary<TAsset, TAssetEntry> oldAssets = (assets ?? Array.Empty<TAssetEntry>()).Where(a => a.Asset != null).ToDictionary(a => a.Asset);
-
-            HashSet<TAsset> unvisitedAssets = new HashSet<TAsset>(assets == null ? Enumerable.Empty<TAsset>() : assets.Where(a => a.Asset != null).Select(a => a.Asset));
+            Dictionary<string, List<AssetId>> oldAssets = lookupByName == null ? new Dictionary<string, List<AssetId>>() : lookupByName;
+            HashSet<Tuple<string, AssetId>> unvisitedIds = new HashSet<Tuple<string, AssetId>>();
+            lookupByAsset = new Dictionary<TAsset, AssetCacheEntry>();
+            lookupByAssetId = new Dictionary<AssetId, AssetCacheEntry>();
+            foreach (var listPair in oldAssets)
+            {
+                foreach (var id in listPair.Value)
+                {
+                    unvisitedIds.Add(new Tuple<string, AssetId>(listPair.Key, id));
+                }
+            }

```

Comment:
This is a medium-sized function that has non-trivial data-reconciliation steps - usually when I see these sorts of things, it's helpful if there's some comments/overall section that explains the overall algorithm that's running through to set the expectations for what the code is doing.

The ask here isn't document line-by-line what is happening (i.e. it's obvious that we're visiting old assets and adding them to a hashset for reconciliation), but more an overall set of steps.